### PR TITLE
Add a convenience function for floating-point comparisons

### PR DESCRIPTION
**Pull request by kalekundert**
This pull request adds a convenience class to assert that two floating-point numbers (or two sets of numbers) are equal to each other within some margin.  You give the class the number you're expecting to get, and the class overrides the "==" operator to do the comparison behind the scenes with a reasonable margin for error.    Most unit testing frameworks have some sort of  ``assertAlmostEqual()`` function to provide this functionality, but to my knowledge ``pytest`` doesn't have anything of the sort.  I think this pull request provides a useful feature in a way that dovetails nicely with the way ``pytest`` deconstructs assert statements.

Tests and documentation are included.  I ran all the tests on python2.7 and python3.4, but I'm a little worried that the way I use unicode in ``approx.__repr__`` might not work in python3.2.  I copied most of the new documentation below for convenience:

Due to the intricacies of floating-point arithmetic, numbers that we would intuitively expect to be the same are not always so:

    >>> 0.1 + 0.2 == 0.3
    False

This problem is commonly encountered when writing tests, e.g. when making sure that floating-point values are what you expect them to be.  One way to deal with this problem is to assert that two floating-point numbers are equal to within some appropriate margin:

    >>> abs((0.1 + 0.2) - 0.3) < 1e-6
    True
   
However, comparisons like this are tedious to write and difficult to understand.  Furthermore, absolute comparisons like the one above are usually discouraged in favor of relative comparisons, which can't even be easily written on one line.  The ``approx`` class provides a way to make floating-point comparisons that solves both these problems:

    >>> from pytest import approx
    >>> 0.1 + 0.2 == approx(0.3)
    True

``approx`` also makes is easy to compare ordered sets of numbers, which would otherwise be very tedious:

    >>> (0.1 + 0.2, 0.2 + 0.4) == approx((0.3, 0.6))
    True

By default, ``approx`` considers two numbers to be equal if the relative error between them is less than one part in a million (e.g. ``1e-6``).  Relative error is defined as ``abs(x - a) / x`` where ``x`` is the value you're expecting and ``a`` is the value you're comparing to.  This definition breaks down when the numbers being compared get very close to zero, so ``approx`` will also consider two numbers to be equal if the absolute difference between them is less than one part in a trillion (e.g.  ``1e-12``).

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
Overall nice work, ill take a deeper look tomorow

**Comment by The-Compiler** [comment]
Some random thoughts:

- `approx` is described as a function once, and as a class once
- How does this compare to Python 3.5's `math.isclose`? If it doesn't use the same algorithm, maybe it should? I know a lot of thought and discussion went into that one :wink:
- Maybe it should have some more unittests (maybe inspired by the ones of `math.isclose`)?

**Comment by nicoddemus** [comment]
Nice work, thanks for the PR! :smile:

File: _pytest/python.py [hunk 1]
```diff
@@ -1336,6 +1339,112 @@ def __exit__(self, *tp):
...
+    """
+
+    def __init__(self, expected, rel=None, abs=None):
+        self.expected = expected
+        self.max_relative_error = rel
+        self.max_absolute_error = abs
+
+    def __repr__(self):
+        from collections import Iterable
+        utf_8 = lambda s: s.encode('utf-8') if sys.version_info.major == 2 else s
```
Comment (nicoddemus) [inline _pytest/python.py]: Unfortunately this is not supported in py26: `sys.version_info.major`. You will have to use `sys.version_info[0] == 2`.

**Comment by tadeu** [comment]
:+1:

**Comment by kalekundert** [comment]
@The-Compiler Thanks for pointing out ``math.isclose``, I didn't know about it.  I did know about ``numpy.isclose``, but for some reason I wasn't thinking about it when I wrote this code.  So I just spent a few minutes comparing my class to those two functions.  The basic algorithm is the same in all the cases, but there are a few meaningful differences:

##### ``Inf`` and ``NaN``:

The standard library and numpy both handle these cases properly, but I don't.  So I should fix that.

##### Default tolerances:

- ``math.isclose(rel_tol=1e-9, abs_tol=0.0)``
- ``numpy.isclose(rtol=1e-5, atol=1e-8)``
- ``pytest.approx(rel=1e-6, abs=1e-12)``

Seeing as how the standard library and numpy don't agree on the defaults, it's probably not that important.  But I don't think it's a good idea for the default absolute tolerance to be ``0.0``.  This means that  ``math.isclose(x, 0.0)`` is the same as ``x == 0.0`` by default, which strikes me as something that could be unpleasantly surprising.  That behavior might be more correct in some mathematical sense, but it's probably less useful for testing.  I could use the numpy defaults, which are a little bit more relaxed than the ones I made up.

##### Application of tolerances:

- ``math.isclose``: True if the relative tolerance is met w.r.t. either ``a`` or ``b`` or if the absolute tolerance is met.
- ``numpy.isclose``: Adds the relative tolerance w.r.t. ``b`` and the absolute tolerance.  Asymmetric, because it doesn't consider the relative tolerance w.r.t. ``a``.
- ``pytest.approx``: True if the relative tolerance is met w.r.t ``b`` or is the absolute tolerance is met.  Asymmetric, but I think that makes sense for testing where ``b`` is the "expected value".  In the special case that the user explicitly specifies an absolute tolerance but not a relative tolerance, only the absolute tolerance is considered.

I don't think these distinctions are really important, because you won't usually be close enough to the tolerance to notice the difference.  That said, I like the way I did it the best. (I guess that's not surprising.)  I like that if the user explicitly specifies an absolute tolerance, then that's exactly what gets used.

Overall I think I'll incorporate some ideas from the standard library and numpy, but I won't provide the exact same interface as either.  I'll also steal some unit tests from the standard library, like you suggested, because they are much more thorough than the doctests I wrote.

**Comment by The-Compiler** [comment]
Thanks for the explanations! I don't have any hard feelings either way, but it's good to know how it relates to the "previous art" :smile:

**Comment by RonnyPfannschmidt** [comment]
good research, i think its important to document those differences in a comprehensible manner

**Comment by nicoddemus** [comment]
Excellent summary! I agree with @RonnyPfannschmidt, I would strongly encourage you to add it to the documentation.

**Comment by kalekundert** [comment]
I think this branch is ready to merge.  Let me know what you think.

File: CHANGELOG.rst [hunk 2]
```diff
@@ -7,7 +7,8 @@
   namespace in which your doctests run.
   Thanks `@milliams`_ for the complete PR (`#1428`_).
 
-* 
+* New ``approx()`` function for easily comparing floating-point numbers in 
+  tests.
```
Comment (nicoddemus) [inline CHANGELOG.rst]: Please add a `"Thanks @kalekundert for the complete PR"`. :smile:

**Comment by nicoddemus** [comment]
Looks very good to me! :grin:

**Comment by nicoddemus** [comment]
I agree it seems ready to merge, after others take another look!

File: _pytest/python.py [hunk 3]
```diff
@@ -1336,6 +1340,253 @@ def __exit__(self, *tp):
...
+    def __init__(self, expected, rel=None, abs=None):
+        self.expected = expected
+        self.abs = abs
+        self.rel = rel
+
+    def __repr__(self):
+        return ', '.join(repr(x) for x in self.expected)
+
+    def __eq__(self, actual):
+        from collections import Iterable
```
Comment (The-Compiler) [inline _pytest/python.py]: Why not do `import collections` at the top of the file instead?

File: _pytest/python.py [hunk 4]
```diff
@@ -1336,6 +1340,253 @@ def __exit__(self, *tp):
...
+        if not isinstance(actual, Iterable): actual = [actual]
+        if len(actual) != len(self.expected): return False
```
Comment (The-Compiler) [inline _pytest/python.py]: I'd prefer to have those conditionals on two separate lines - having `if foo: return False` on one line just makes things harder to read and easier to miss the `return`, IMHO.

File: _pytest/python.py [hunk 5]
```diff
@@ -1336,6 +1340,253 @@ def __exit__(self, *tp):
...
+            return str(self.expected)
+
+        # If a sensible tolerance can't be calculated, self.tolerance will
+        # raise a ValueError.  In this case, display '???'.
+        try:
+            vetted_tolerance = '{:.1e}'.format(self.tolerance)
+        except ValueError:
+            vetted_tolerance = '???'
+
+        repr = u'{0} \u00b1 {1}'.format(self.expected, vetted_tolerance)
```
Comment (The-Compiler) [inline _pytest/python.py]: You're shadowing the builtin `repr` here - while it's only locally and doesn't matter, I'd still prefer you didn't :wink:

File: testing/python/approx.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+    def __init__(self):
+        doctest.DocTestRunner.__init__(self)
+
+    def report_failure(self, out, test, example, got):
+        raise AssertionError("'{}' evaluates to '{}', not '{}'".format(
+            example.source.strip(), got.strip(), example.want.strip()))
+
+
+class TestApprox:
```
Comment (The-Compiler) [inline testing/python/approx.py]: You could use [`@pytest.mark.parametrize`](http://pytest.org/latest/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions) here to separate tests from each other (so e.g. other tests still run if one of them fails), but I'm okay with the current form as well.

**Comment by The-Compiler** [comment]
Thanks! I added some more line notes for stuff I noticed. Great work overall!

I'm also wondering if this should live in the `pytest` namespace, as we might want to add other similar "comparison helpers" later, see #95. Maybe it's time to add a `pytest.helpers` namespace instead? I don't have any hard feelings on this either way though. What do you think?

**Comment by RonnyPfannschmidt** [comment]
lets document it as experimental - its not hard to provide deprecation helpers later on to support a rename

**Comment by nicoddemus** [comment]
Great comments @The-Compiler overall.

I don't think we need to introduce the `pytest.helpers` change right now... TBH I don't think the `pytest` namespace is bloated as it is now. I also don't want to burden @kalekundert any longer, as I think the PR is outstanding work as it is. :grin:

**Comment by RonnyPfannschmidt** [comment]
unfortunately my previous web ui merge created a conflict, i can merge manually after finishing setting up my pc, but if anyone beats me to it, feel free, just make an note when you start

**Comment by kalekundert** [comment]
I'll fix the merge conflict.

For what it's worth regarding ``pytest.helpers``, I prefer package hierarchies that are as flat as possible.  It's less to type and less to remember, and as long as everything still has a clear name, its doesn't cause any problems.  I also think that if you do make a ``pytest.helpers`` namespace, then you should also move ``pytest.raises`` into it.

**Comment by RonnyPfannschmidt** [comment]
Well done, thanks for the continued good work

**Comment by bilderbuchi** [comment]
Thank you for this work, at last there is approximate comparison in pytest (so far I have always used `numpy.testing.assert_almost_equal` and friends).

It's great that in the docstring the different methods of approximate comparison are listed and discussed! However, in light of this I find it rather confusing that `pytest.approx` is the *only* function that has a totally different API/usage pattern - they all have  `function(a,b,...)`, only approx has `a == approx(b)`. This is unfortunate in my eye. Was there a technical reason to deviate from the "standard" way to do comparisons in this case?

**Pull request merged**